Jenkinsfile

```diff
@@ -0,0 +1,47 @@
+pipeline {
+    agent any
+    parameters {
+        choice(name: 'DEPLOY_TO', choices: "\ntest\nacceptance\nproduction\ninternal", description: 'Which CF space to deploy to.')
+        string(name: 'DOCKER_PORT', defaultValue: '49767', description: 'HTTP port on which Docker node image should listen.')
+    }
+    stages {
+        stage('Build') {
+            agent {
+                dockerfile {
+                    args "-p ${params.DOCKER_PORT}:3000"
+                }
+            }
+            environment {
+                HOME = '.'
+            }
+            steps {
+                checkout scm
+                sh 'npm install'
+                sh 'npm run build'
+            }
+        }
+        stage('Deploy to CF test space') {
+            when {
+                anyOf {
+                    branch 'develop'
+                    environment name: 'DEPLOY_TO', value: 'test'
+                }
+            }
+            environment {
+                CF_HOME="${env.HOME}/.cf/${env.JOB_NAME}"
+                CF_API="${env.CF_API}"
+                CF_LOGIN=credentials('portaljs.cloudfoundry.login')
+                CF_ORG="${env.CF_ORG}"
+                CF_SPACE="test"
+                CF_APP_NAME="portaljs-test"
+            }
+            steps {
+                sh 'mkdir -p ${CF_HOME}'
```

Comment:
The `CF_HOME` environment variable, if set, customises where CF CLI configuration is stored. This is essential to set per-job on Jenkins to ensure that multiple concurrent jobs do not interfere with each other.

For instance:
1. Jobs A and B are started at roughly the same time, without `CF_HOME` customised, so defaulting to the jenkins user's `~/.cf` directory
2. Job A runs `cf target -s test`
3. Job B runs `cf target -s production`
4. Job A runs `cf push` but instead of pushing to test this pushes to production

See https://docs.cloudfoundry.org/cf-cli/getting-started.html#login